common/hdr-model.cpp

```diff
@@ -0,0 +1,345 @@
+            if( ctrls[0].contains( "depth-gain" ) )
+                ctrl.DepthGain = std::stoi( ctrls[0]["depth-gain"].get< std::string >() );
+            if( ctrls[1].contains( "depth-exposure" ) )
+                ctrl.DepthManExp = std::stoi( ctrls[1]["depth-exposure"].get< std::string >() );
+
+            item.controls.push_back( ctrl );
+        }
+        SubPreset.items.push_back( std::move( item ) );
+    }
+}
+
+hdr_model::hdr_model( rs2::device dev )
+    : _device( dev )
+    , _window_open( false )
+    , _hdr_supported( false )
+    , _set_default( false )
+{
+    _hdr_supported = check_HDR_support();
+    if( _hdr_supported )
+    {
+        initialize_default_config();
+        _current_config = _default_config;
+        _changed_config = _current_config;
+    }
+}
+
+bool hdr_model::check_HDR_support()
+{
+    try
+    {
+        if( ! _device.is< rs2::serializable_device >() )
+            return false;
+        auto serializable = _device.as< rs2::serializable_device >();
+        return ! serializable.serialize_json().empty();
+    }
+    catch( ... )
+    {
+    }
+    return false;
+}
+
+void hdr_model::initialize_default_config()
+{
+    _default_config.SubPreset.id = "0";
+    _default_config.SubPreset.iterations = 0;
+
+    const std::vector< std::pair< int, int > > defaults
+        = { { 1, 1 }, { 32, 1000 }, { 128, 100000 }, { 2, 2 }, { 232, 200 }, { 2128, 20000 } };
+    for( auto const & p : defaults )
+    {
+        hdr_config::preset_item it;
+        it.iterations = 1;
+        it.controls.emplace_back( p.first, p.second );
+        _default_config.SubPreset.items.push_back( it );
+    }
+}
+
+bool hdr_model::supports_HDR() const
+{
+    return _hdr_supported;
+}
+
+void hdr_model::load_hdr_config_from_file( const std::string & filename )
+{
+    std::ifstream file( filename );
+    if( ! file )
+        throw std::runtime_error( "Cannot open file " + filename );
+    std::string content( ( std::istreambuf_iterator< char >( file ) ), {} );
+    _changed_config.from_json( content );
+}
+
+void hdr_model::save_hdr_config_to_file( const std::string & filename )
+{
+    std::ofstream file( filename );
+    if( ! file )
+        throw std::runtime_error( "Cannot write file " + filename );
+    file << _current_config.to_json();
+}
+
+void hdr_model::apply_hdr_config()
+{
+    if( ! _device.is< rs2::serializable_device >() )
+        throw std::runtime_error( "Device not serializable" );
+    auto serializable = _device.as< rs2::serializable_device >();
+    serializable.load_json( _changed_config.to_json() );
+    _current_config = _changed_config;
+}
+
+void hdr_model::render_control_item( hdr_config::control_item & control, int /*idx*/ )
+{
+    ImGui::PushID( &control );
+
+    ImGui::Text( "Depth Gain:" );
+    ImGui::SameLine();
+    ImGui::SetNextItemWidth( 100 );
+    int g = control.DepthGain;
+    if( ImGui::InputInt( "##gain", &g ) )
+        control.DepthGain = std::max( 1, std::min( g, 16384 ) );
```

Comment:
For D455 (and other models) gain range is 16-248. Can we get values from actual option?